### DISCUSSION TITLE
Inquiry about abnormal vibration after launching CRTK topics

### DISCUSSION THREAD

**Discussion by jackzhy96** [surgical-robotics-ai/surgical_robotics_challenge#30] [Q&A] [answered]
Hi Adnan,

Hope all is well.

I currently come across a question that the robot has abnormal vibration after running launch_CRTK script.

To be specific, please refer to my repo: https://github.com/JackHaoyingZhou/surgical_robotics_challenge 

I collected some data of task2 or 3 using Geomagic haptic device. You can run direct_multi_psm_control.py  and modify the file path at https://github.com/JackHaoyingZhou/surgical_robotics_challenge/blob/devel/scripts/surgical_robotics_challenge/teleoperation/direct_multi_psm_control.py#L73

If I run without launching CRTK topics, all movements go very smoothly and can accomplish the task. Nevertheless, if I launch the CRTK topics, it will vibration even though I have **modified the "add_joint_errors" option to False** at https://github.com/JackHaoyingZhou/surgical_robotics_challenge/blob/devel/scripts/surgical_robotics_challenge/launch_crtk_interface.py#L107

In addition, I also modified the publish rate but it didn't work. Would you mind letting me know how I can modify so that the vibration can disappear? 

Thanks!

Best,
Haoying(Jack) Zhou

**Comment by adnanmunawar** [comment]
Hi Jack,

Thanks for bringing this to our attention. Can you record a short clip of both modes (i.e. with and without running launch_crtk interface)? Essentially, I want to see how bad is the vibration and what is the physics frequency. The physics frequency is displayed on the simulator window or it can be collected via the `/ambf/env/World/State` topic. If possible, can you plot it out for better visualization? And before you do these, can you also update your local AMBF to the latest version?

**Reply by jackzhy96** [reply]
Hi Adnan,

I have followed your advice and updated my local AMBF to the latest version. Yo can find corresponding videos via following link:

Launching CRTK:  https://youtu.be/Phuia7m8TOk
Not launching CRTK: https://youtu.be/hR7xD3lqOZo

In addition, if I stop running launch_CRTK script, the vibrations are gone. Other than the vibrations, it seems that it also has some misalignment when launching CRTK. 

Please let me know if you need any additional information, thanks! Look forward to hearing from you!

Best,
Jack

**Comment by adnanmunawar** [comment, marked as answer]
Hi Jack, thank you for creating the nice videos. I noticed that you are using the `launch_crtk_interface.py` and `direct_multi_psm_control.py` together. Then I went to your [`direct_multi_psm_control.py`](https://github.com/JackHaoyingZhou/surgical_robotics_challenge/blob/devel/scripts/surgical_robotics_challenge/teleoperation/direct_multi_psm_control.py#L120) and saw that it uses the methods `servo_cp` which calls `servo_jp` to directly control the PSM joint angles. What I believe is happening is that there is infighting going on between the `direct_multi_psm_control.py` and `launch_crtk_interface.py` as it also directly controls the PSM and ECM joint angles using `servo_jp`. 

In conclusion, when using `launch_crtk_interface.py`, you should only ever control the PSMs / ECMs using the CRTK ROS topics. Does this make sense? If you want to continue using the `direct_multi_psm_control.py`, you should change the relevant control commands with ROS publishers that publish on the topics exposed by `launch_crtk_interface.py`.

**Reply by jackzhy96** [reply]
Hi Adnan,

Thanks for your answer! I will try out your advice and get back to you soon.

-Jack
